db-tests/src/db_tests.rs

```diff
@@ -0,0 +1,1079 @@
+    coll: &str,
+    bid: &str,
+    payload: Option<&str>,
+    sortindex: Option<i32>,
+    ttl: Option<u32>,
+) -> params::PutBso {
+    params::PutBso {
+        user_id: HawkIdentifier::new_legacy(u64::from(user_id)),
+        collection: coll.to_owned(),
+        id: bid.to_owned(),
+        payload: payload.map(|payload| payload.to_owned()),
+        sortindex,
+        ttl,
+    }
+}
+
+fn postbso(
+    bid: &str,
+    payload: Option<&str>,
+    sortindex: Option<i32>,
+    ttl: Option<u32>,
+) -> params::PostCollectionBso {
+    params::PostCollectionBso {
+        id: bid.to_owned(),
+        payload: payload.map(&str::to_owned),
+        sortindex,
+        ttl,
+    }
+}
+
+fn gbso(user_id: u32, coll: &str, bid: &str) -> params::GetBso {
+    params::GetBso {
+        user_id: hid(user_id),
+        collection: coll.to_owned(),
+        id: bid.to_owned(),
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+fn gbsos(
+    user_id: u32,
+    coll: &str,
+    bids: &[&str],
+    older: u64,
+    newer: u64,
+    sort: Sorting,
+    limit: i64,
+    offset: i64,
+) -> params::GetBsos {
+    params::GetBsos {
+        user_id: hid(user_id),
+        collection: coll.to_owned(),
+        params: BsoQueryParams {
+            ids: bids.iter().map(|id| id.to_owned().into()).collect(),
+            older: Some(SyncTimestamp::from_milliseconds(older)),
+            newer: Some(SyncTimestamp::from_milliseconds(newer)),
+            sort,
+            limit: Some(limit as u32),
+            offset: Some(offset as u64),
+            full: true,
+        },
+    }
+}
+
+fn dbso(user_id: u32, coll: &str, bid: &str) -> params::DeleteBso {
+    params::DeleteBso {
+        user_id: hid(user_id),
+        collection: coll.to_owned(),
+        id: bid.to_owned(),
+    }
+}
+
+fn dbsos(user_id: u32, coll: &str, bids: &[&str]) -> params::DeleteBsos {
+    params::DeleteBsos {
+        user_id: hid(user_id),
+        collection: coll.to_owned(),
+        ids: bids.iter().map(|id| id.to_owned().into()).collect(),
+    }
+}
+
+fn hid(user_id: u32) -> HawkIdentifier {
+    HawkIdentifier::new_legacy(u64::from(user_id))
+}
+
+macro_rules! with_delta {
+    ($db:expr, $delta:expr, $body:block) => {{
+        let ts = $db.timestamp().as_i64();
+        $db.set_timestamp(SyncTimestamp::_from_i64(ts + $delta).unwrap());
+        let result = $body;
+        $db.set_timestamp(SyncTimestamp::_from_i64(ts).unwrap());
+        result
+    }};
+}
+
+#[async_test]
+async fn bso_successfully_updates_single_values() -> Result<()> {
+    let db = db().await?;
+
+    let uid = 1;
+    let coll = "clients";
+    let bid = "testBSO";
+    let sortindex = 1;
+    let ttl = 3600 * 1000;
+    let bso1 = pbso(
+        uid,
+        coll,
+        bid,
+        Some("initial value"),
+        Some(sortindex),
+        Some(ttl),
+    );
+    db.put_bso(bso1).compat().await?;
+
+    let payload = "Updated payload";
+    let bso2 = pbso(uid, coll, bid, Some(payload), None, None);
```

Comment:
nit; purely for readability, but I wonder if enclosing these sub tests in blocks might be a good idea. It would help with possible cross contamination.